Locus Explorer
=============
An interactive graphical illustration of genomic regions of interest.

### Disclaimer
Locus Explorer is currently under development and may include errors in both plots and analyses. Any results provided by Locus Explorer should be used with caution.
**Next update is planned for 31 July 2015.**

### Availability  
The source code and install instructions for Locus Explorer are available at https://github.com/oncogenetics/LocusExplorer.

### Run
```R
require(shiny)  
runGitHub("LocusExplorer", "oncogenetics")
```

### Cite
[Multiple novel prostate cancer susceptibility signals identified by fine-mapping of known risk loci among Europeans.](http://www.ncbi.nlm.nih.gov/pubmed/26025378) Al Olama AA<sup>1</sup>, Dadaev T<sup>2</sup>
##### Abstract
Genome-wide association studies (GWAS) have identified numerous common prostate cancer (PrCa) susceptibility loci. We have fine-mapped 64 GWAS regions known at the conclusion of the iCOGS study using large scale genotyping and imputation in 25,723 PrCa cases and 26,274 controls of European ancestry.We detected evidence for multiple independent signals at 16 regions, 12 of which contained additional newly identified significant associations. A single signal comprising a spectrum of correlated variation was observed at 39 regions; 35 of which are now described by a novel more significantly associated lead SNP, whilst the originally reported variant remained as the lead SNP only in 4 regions. We also confirmed 2 association signals in Europeans that had been previously reported only in East-Asian GWAS.Based on statistical evidence and LD structure we have curated and narrowed down the list of the most likely candidate causal variants for each region. Functional annotation using data from ENCODE filtered for PrCa cell lines and eQTL analysis demonstrated significant enrichment for overlap with bio-features within this set. By incorporating the novel risk variants identified here alongside the refined data for existing association signals, we estimate that these loci now explain approximately 38.9% of the familial relative risk of PrCa, an 8.9% improvement over the previously reported GWAS tag SNPs. This suggests that a significant fraction of the heritability of PrCa may have been hidden during the discovery phase of GWAS, in particular due to the presence of multiple independent signals within the same region.

### Contact  
Questions, suggestions, and bug reports are welcome and appreciated. Please contact Tokhir Dadaev at tokhir.dadaev@icr.ac.uk.

